### DISCUSSION TITLE
Trim without resetting to initial conditions

### DISCUSSION THREAD

**Discussion by ThomasKennings** [JSBSim-Team/jsbsim#649] [Q&A]
Edit: Marked a response as an answer. To future readers: I'm planning to use the PID feature described in section 3.1.9.1.17 of the JSBSim PDF manual to hold attitudes, etc. It's easy enough to implement (just a few lines in the aircraft xml file).

Hello,

I am trying to write a script such that the 737 enters a bank of a given angle, and then sets do_simple_trim = 5 to hold the turn steadily. However, setting the trim mode once banked results in the aircraft reverting to initial conditions. Is there a way to set the trim without re-initializing the aircraft? Is do_simple_trim the right tool for this job, or am I going about the problem of a stable turn completely wrong? My end goal is to be able to turn the plane in, say, a holding pattern. 

Github won't let me attach my XML, so...
```xml
<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet type="text/xsl" href="http://jsbsim.sf.net/JSBSimScript.xsl"?>
<runscript xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:noNamespaceSchemaLocation="http://jsbsim.sf.net/JSBSimScript.xsd"
    name="Left turn in a 737.">

  <use aircraft="737" initialize="90_degree_left_turn_init"/> 

  <run start="0" end="100" dt="0.008333">

    <!--
      For "do_simple_trim" (Classic trim):
      0: Longitudinal
      1: Full
      2: Ground
      3: Pullup
      4: Custom
      5: Turn
      6: None
    -->
    
    <event name="Trim Level">
      <condition>
        simulation/sim-time-sec ge 0
      </condition>
      <set name="simulation/do_simple_trim" value="1"/>
    </event>

    <event name="Remove Level Trim">
      <condition>
        simulation/sim-time-sec ge 10
      </condition>
      <set name="simulation/do_simple_trim" value="6"/>
    </event>

    <event name="Enter Left Bank" continuous="true">
      <condition>
        simulation/sim-time-sec ge 10
      </condition>
      <set name="fcs/aileron-cmd-norm">
	<function>
	<table>
		<independentVar lookup="row">attitude/phi-deg</independentVar>
		<tableData>
		  0.0	0.2
		  10.0	0.2
		  15.0	0.1
		  20.0	0.0
		</tableData>
	</table>
	</function>
      </set>
      <notify>
        <property>simulation/do_simple_trim</property>
      </notify>
    </event>

    <event name="Trim Turn">
      <condition>
        attitude/phi-deg ge 19
      </condition>
      <set name="simulation/do_simple_trim" value="5"/>
    </event>

  </run>

</runscript>
```

**Comment by dpculp** [comment]
It looks like you're using the trim system to do what would normally be done by an autopilot.  I've not seen the trim system used in this way, and I'm not surprised it's resetting your airplane to initial conditions.  Maybe there's a way to force it to work, but I doubt it.

Have you tried the autopilot approach?

**Reply by ThomasKennings** [reply]
My goal is to test a revised version of the MCAS module, which was part of the fly-by-wire system in the 737-MAX. It was implemented so that pilots would be able to fly the 737-MAX just like the 737-NG. I am trying to replicate flight of the aircraft by the pilot, and am concerned that using autopilot would go against this.

I am considering writing a continuous event that will maintain pitch angle during the turn, and maybe an event for coordinating the turns.  Do you have any thoughts on this? I may play with scripting autopilot to see how it controls the aircraft (does its control go through the piltos controls, etc.).

**Reply by dpculp** [reply]
You can trim the aircraft to fly the turn, but once the trimming is over the airplane is unpiloted, so it will eventually enter a spiral.  Without a pilot in the loop the only way to maintain the maneuver is with an autopilot.

On the other hand, if you only need to hold the turn for a few seconds and log some data, then you can get away with no pilot (human or auto), but remember you don't have a way to introduce reasonable control inputs.  You can script some control inputs, but the script won't have knowledge of the aircraft's state, so they won't necessarily be reasonable inputs.

**Comment by seanmcleod** [comment]
>  fly-by-wire system in the 737-MAX

The 737 series isn't really a FBW aircraft. The Airbus A320 was the first commercial airliner to include FBW functionality. Stability augmentation systems, stick pushers near the stall, MAX re-trimming the horizontal stabilizer etc. don't make it a FBW aircraft.

> continuous event that will maintain pitch angle during the turn

Now it sounds like you're sort-of adding FBW like functionality, e.g. the Airbus FBW aircraft do exactly this for bank angles < 33 degrees.

Not really clear what you're trying to do. Are you trying to generate some automated flight tests, e.g. perform a climbing turn with increasing alpha until MCAS kicks in, but with your own modified MCAS algorithm?

**Reply by ThomasKennings** [reply]
Ah, my mistake with the FBW. I am working on a project that involves comparing several MCAS modules against one another to see differences in characteristics, and am trying to make repeatable flights so they can be compared fairly.

**Comment by seanmcleod** [comment]
If you give us more details on the exact types of flights you're trying to repeatably run against different MCAS implementations then we could potentially give you some useful pointers.

As @dpculp mentioned:

> On the other hand, if you only need to hold the turn for a few seconds and log some data, then you can get away with no pilot (human or auto)

So for example if your test is something like trimmed for landing, then executing a go-around at 100ft, e.g. with full thrust input and say 25% elevator input applied and held and then seeing how quickly that gets you to a stall angle with no MCAS, with MCAS implementation 1, 2, 3 etc. then you can script that quite easily. 

> but remember you don't have a way to introduce reasonable control inputs. You can script some control inputs, but the script won't have knowledge of the aircraft's state, so they won't necessarily be reasonable inputs.

So in my example as Dave mentioned, there is no change to the inputs based on the pilot's view of the aircraft's current state.

The next level is some form of auto-pilot/PID control, e.g. maybe you want to simulate a pilot getting fixated on say holding a specific pitch attitude of say 15 degrees while not applying enough go-around power and then comparing what would happen with no MCAS, with MCAS implementation 1, 2, 3 etc. as alpha increases etc.

**Reply by ThomasKennings** [reply]
Here is a list of the tests I'm trying to make, roughly in order of complexity:
1. Straight and level flight (I was able to make this one simply using do_simple_trim=1)
2. Level flight into a series of turns left and right
3. Level flight into an altitude change, into level flight again
4. Holding pattern, perhaps with an altitude change at some point
5. Take-off
6. Landing

PID to simulate pilot behavior/error would be excellent, though the scripting doesn't seem to support this too easily based on what I've used so far. Auto-pilot was my first idea for executing some of these maneuvers, but I'm hesitant about how, or whether, to interface our MCAS with the autopilot.

You guys seem like the experts on this--what are your thoughts? Thank you, BTW, for your advice so far!

**Comment by seanmcleod** [comment]
And under what conditions is the MCAS going to be triggered? These all seem to be pretty benign manoeuvres.  

PID controllers aren't implemented in the script, they're implemented in the aircraft xml file or a separate systems xml file. But they can expose properties like set-points, boolean properties to enable/disable themselves etc. which can be easily set in the script.

**Reply by ThomasKennings** [reply]
That's undecided right now. The MCAS is in early development currently.

I just did another search through the documentation for PID controls and found a whole section 2.7 of the PDF manual. This looks like a promising tool! I'm going to give this a shot and see if I can get it to work, and see if it will meet my needs.

I agree that the maneuvers seem like they should be scriptable. I might've been missing this gigantic piece of the puzzle. I will report with how it goes.

**Comment by ThomasKennings** [comment]
I see two implementations of PID in the manual: one described in section 2.7, and one described in section 3.1.9.1.17. Which is more appropriate for trying to simulate a pilot's efforts to keep the plane on a given course (ex: level flight)? The implementation in 3.1.9.1.17 seems far simpler.

**Comment by seanmcleod** [comment]
Section 2.7 implements a PID controller using multiple separate elements like `pure_gain`, `integrator` and `summer`. As opposed to making use of the single `pid` element described in 3.1.9.1.17. I'm not sure of the history, maybe when section 2.7 was written there was no `pid` element, possibly only added at a later stage.

So yes I'd go with making use of the `pid` element.

**Reply by ThomasKennings** [reply]
Thanks for clarifying the history! I did get the pid element working this morning for controlling the roll of the airplane. I then tried to do the same for the pitch, but things got weird. The PID is working perfectly, but the elevator command is not changing the angle of the elevator. The image below shows this (and shows the success of the roll PID). Does anyone have advice about why the elevator command isn't actuating the elevator? The Elevator command is clipped -1 to 1 since it's a normalized value. And I know it's an extreme gain on the elevator PID but I'll tune that once I get it working :P
![Elevator Problem](https://user-images.githubusercontent.com/64053736/173199158-cbea8e8b-7fdc-4b2e-9d3d-a20b0dde4b4d.jpg)

And here's the PID element for elevator. Aileron control looks similar.
```xml	
<channel name="Elevator PID">
<pid name="fcs/elevator-cmd-norm">
<input>attitude/theta-deg</input>
<kp> 1.0 </kp>
<clipto>
<min> -1.0 </min>
<max> 1.0 </max>
</clipto>
</pid>
</channel>
````

**Reply by bcoconni** [reply]
FWIW the `<max>` subelement of `<clipto>` is missing. It should be:
```xml
<clipto>
  <min> -1.0 </min>
  <max> 1.0 </max>
</clipto>
```

**Reply by ThomasKennings** [reply]
I checked my xml file and its there. I fixed my post. I probably made an error while copy-pasting.

**Comment by seanmcleod** [comment]
> is not changing the angle of the elevator.

But the **Elevator Position (deg)** graph shows the elevator angle changing during the 30s period of the graph.

**Reply by ThomasKennings** [reply]
I would imagine that 100% pilot input would have a greater, and much different looking, impact than what is shown on the graphs. The graphs do not change if I set gain = 0. I suspect there is something else operating behind the scenes on my end. I and my team have done a bit of modding, so maybe it was that. In any case, I'm going to mark this thread as answered since the PID should take care of my original issue! Appreciate the help! (I may be back soon lol).
